File: src/rust/src/summary.rs
#![allow(dead_code)]

use serde::{Deserialize, Serialize};

use crate::output::is_dark;
use crate::state::GameState;
use crate::types::{INV_LOCATION, MAX_ITEMS};

#[derive(Serialize, Deserialize, Default, Clone)]
pub struct GameSummary {
    pub room_id: usize,
    pub room_name: String,
    pub turns: i32,
    pub score: i32,
    pub is_playing: bool,
    pub is_riding: bool,
    pub is_dark: bool,
    pub thirst: i32,
    pub horse_thirst: i32,
    pub has_water: bool,
    pub lamp_lit: bool,
    pub horse_saddled: bool,
    pub inventory: Vec<String>,
}

pub fn summarize_state(s: &GameState) -> GameSummary {
    let mut summary = GameSummary {
        room_id: s.current_room,
        room_name: s.rooms[s.current_room]
            .as_ref()
            .map(|r| r.name.clone())
            .unwrap_or_default(),
        turns: s.turns,
        score: s.score,
        is_playing: s.is_playing,
        is_riding: s.is_riding,
        is_dark: is_dark(s),
        thirst: s.thirst,
        horse_thirst: s.horse_thirst,
        has_water: s.has_water,
        lamp_lit: s.is_lamp_lit,
        horse_saddled: s.is_horse_saddled,
        inventory: Vec::new(),
    };
    for i in 1..=MAX_ITEMS {
        if s.items[i].location == INV_LOCATION {
            summary.inventory.push(s.items[i].description.clone());
        }
    }
    summary
}

pub fn describe_room(s: &GameState) -> String {
    let Some(room) = s.rooms[s.current_room].as_ref() else {
        return "You are in an unknown location.".to_string();
    };

    let mut desc = format!("{}\n{}\n", room.name, room.description);

    let items: Vec<&str> = (1..=MAX_ITEMS)
        .filter(|&i| s.items[i].location == s.current_room as i32)
        .map(|i| s.items[i].description.as_str())
        .collect();
    if !items.is_empty() {
        desc.push_str(&format!("Items: {}\n", items.join(", ")));
    }

    let mut exits = Vec::new();
    if room.north > 0 { exits.push("north"); }
    if room.south > 0 { exits.push("south"); }
    if room.east  > 0 { exits.push("east"); }
    if room.west  > 0 { exits.push("west"); }
    if !exits.is_empty() {
        desc.push_str(&format!("Exits: {}", exits.join(", ")));
    }

    desc.trim_end().to_string()
}

pub fn describe_inventory(s: &GameState) -> String {
    let items: Vec<&str> = (1..=MAX_ITEMS)
        .filter(|&i| s.items[i].location == INV_LOCATION)
        .map(|i| s.items[i].description.as_str())
        .collect();

    if items.is_empty() {
        "You are carrying nothing.".to_string()
    } else {
        format!("You are carrying: {}", items.join(", "))
    }
}

pub fn summarize_state_json(s: &GameState) -> String {
    serde_json::to_string(&summarize_state(s)).unwrap_or_else(|_| "{}".to_string())
}

File: src/dotnet/Summary.cs
namespace Dustwood;

using System.Text.Json;
using System.Text.Json.Serialization;

public class GameSummary
{
    [JsonPropertyName("room_id")]
    public int RoomID { get; set; }

    [JsonPropertyName("room_name")]
    public string RoomName { get; set; } = "";

    [JsonPropertyName("turns")]
    public int Turns { get; set; }

    [JsonPropertyName("score")]
    public int Score { get; set; }

    [JsonPropertyName("is_playing")]
    public bool IsPlaying { get; set; }

    [JsonPropertyName("is_riding")]
    public bool IsRiding { get; set; }

    [JsonPropertyName("is_dark")]
    public bool IsDark { get; set; }

    [JsonPropertyName("thirst")]
    public int Thirst { get; set; }

    [JsonPropertyName("horse_thirst")]
    public int HorseThirst { get; set; }

    [JsonPropertyName("has_water")]
    public bool HasWater { get; set; }

    [JsonPropertyName("lamp_lit")]
    public bool LampLit { get; set; }

    [JsonPropertyName("horse_saddled")]
    public bool HorseSaddled { get; set; }

    [JsonPropertyName("inventory")]
    public List<string> Inventory { get; set; } = new();
}

public static class SummaryHelper
{
    public static GameSummary SummarizeState(GameState s)
    {
        var summary = new GameSummary
        {
            RoomID = s.CurrentRoom?.Id ?? 0,
            RoomName = s.CurrentRoom?.Name ?? "",
            Turns = s.Turns,
            Score = s.Score,
            IsPlaying = s.IsPlaying,
            IsRiding = s.IsRiding,
            IsDark = IsDark(s),
            Thirst = s.Thirst,
            HorseThirst = s.HorseThirst,
            HasWater = s.HasWater,
            LampLit = s.IsLampLit,
            HorseSaddled = s.IsHorseSaddled
        };

        for (int i = 1; i <= Constants.MaxItems; i++)
        {
            if (s.Items[i].Location == Constants.InvLocation)
            {
                summary.Inventory.Add(s.Items[i].Description);
            }
        }

        return summary;
    }

    public static bool IsDark(GameState s)
    {
        if (s.CurrentRoom == null) return false;
        // Room 4 (Assayer's Office) or Room 7 (Sheriff's Cell) or turn >= DarkTurn
        if (s.CurrentRoom.Id == 4 || s.CurrentRoom.Id == 7)
        {
            return !s.IsLampLit && s.TempLightTurns <= 0;
        }
        if (s.Turns >= Constants.DarkTurn)
        {
            return !s.IsLampLit && s.TempLightTurns <= 0;
        }
        return false;
    }

    public static string DescribeRoom(GameState s)
    {
        if (s.CurrentRoom == null) return "You are in an unknown location.";

        var sb = new System.Text.StringBuilder();
        sb.AppendLine(s.CurrentRoom.Name);
        sb.AppendLine(s.CurrentRoom.Description);

        var items = new List<string>();
        for (int i = 1; i <= Constants.MaxItems; i++)
        {
            if (s.Items[i].Location == s.CurrentRoom.Id)
            {
                items.Add(s.Items[i].Description);
            }
        }
        if (items.Count > 0)
        {
            sb.AppendLine("Items: " + string.Join(", ", items));
        }

        var exits = new List<string>();
        if (s.CurrentRoom.North != null) exits.Add("north");
        if (s.CurrentRoom.South != null) exits.Add("south");
        if (s.CurrentRoom.East != null) exits.Add("east");
        if (s.CurrentRoom.West != null) exits.Add("west");

        if (exits.Count > 0)
        {
            sb.Append("Exits: " + string.Join(", ", exits));
        }

        return sb.ToString().Trim();
    }

    public static string DescribeInventory(GameState s)
    {
        var items = new List<string>();
        for (int i = 1; i <= Constants.MaxItems; i++)
        {
            if (s.Items[i].Location == Constants.InvLocation)
            {
                items.Add(s.Items[i].Description);
            }
        }

        if (items.Count == 0) return "You are carrying nothing.";
        return "You are carrying: " + string.Join(", ", items);
    }

    public static string SummarizeStateJSON(GameState s)
    {
        var summary = SummarizeState(s);
        return JsonSerializer.Serialize(summary);
    }
}
